<h1>PROJECT: United Kingdom Road Accident Data Analysis </h1>
<h2> Inclusive Year: 2019 - 2022 </h2>
<h3>Analyst: Hernandez, Rina Rose</h3>

In [1]:
pip install numpy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.2 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
pip install matplotlib

Defaulting to user installation because normal site-packages is not writeableNote: you may need to restart the kernel to use updated packages.




[notice] A new release of pip is available: 24.2 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
pip install seaborn

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.2 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [4]:
pip install pandas

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.2 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [5]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import f_oneway

<h1> Data Analytics Project </h1>
<h2> United Kingdom Road accident data analysis </h2>
<ul>
    <li>Inlusive years: 2019-2022</li>
</ul>

In [6]:
accident = pd.read_csv('dataset\\accident_data.csv')

In [7]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [8]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [9]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')

In [10]:
accident['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

In [11]:
accident.dtypes

Index                        object
Accident_Severity          category
Accident Date                object
Latitude                    float64
Light_Conditions             object
District Area                object
Longitude                   float64
Number_of_Casualties          int64
Number_of_Vehicles            int64
Road_Surface_Conditions      object
Road_Type                    object
Urban_or_Rural_Area          object
Weather_Conditions           object
Vehicle_Type                 object
dtype: object

In [12]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [13]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mean())

In [14]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [15]:
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mean())

In [16]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [17]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')

In [18]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  object  
 1   Accident_Severity        660679 non-null  category
 2   Accident Date            660679 non-null  object  
 3   Latitude                 660679 non-null  float64 
 4   Light_Conditions         660679 non-null  object  
 5   District Area            660679 non-null  object  
 6   Longitude                660679 non-null  float64 
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  659953 non-null  category
 10  Road_Type                656159 non-null  object  
 11  Urban_or_Rural_Area      660664 non-null  object  
 12  Weather_Conditions       646551 non-null  object  
 13  Vehicle_Type             660679 non-null  ob

In [19]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])

In [20]:
accident.mode()

Index Accident_Severity Accident Date   Latitude Light_Conditions  \
0  2.01E+12            Slight    30/11/2019  52.458798         Daylight   

  District Area  Longitude  Number_of_Casualties  Number_of_Vehicles  \
0    Birmingham  -0.977611                     1                   2   

  Road_Surface_Conditions           Road_Type Urban_or_Rural_Area  \
0                     Dry  Single carriageway               Urban   

   Weather_Conditions Vehicle_Type  
0  Fine no high winds          Car

In [21]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [22]:
accident['Road_Type'] = accident['Road_Type'].astype('category')

In [23]:
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')

In [24]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')

In [25]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  object  
 1   Accident_Severity        660679 non-null  category
 2   Accident Date            660679 non-null  object  
 3   Latitude                 660679 non-null  float64 
 4   Light_Conditions         660679 non-null  object  
 5   District Area            660679 non-null  object  
 6   Longitude                660679 non-null  float64 
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  660679 non-null  category
 10  Road_Type                656159 non-null  category
 11  Urban_or_Rural_Area      660664 non-null  category
 12  Weather_Conditions       646551 non-null  category
 13  Vehicle_Type             660679 non-null  ob

In [26]:
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])

In [27]:
accident.mode()

Index Accident_Severity Accident Date   Latitude Light_Conditions  \
0  2.01E+12            Slight    30/11/2019  52.458798         Daylight   

  District Area  Longitude  Number_of_Casualties  Number_of_Vehicles  \
0    Birmingham  -0.977611                     1                   2   

  Road_Surface_Conditions           Road_Type Urban_or_Rural_Area  \
0                     Dry  Single carriageway               Urban   

   Weather_Conditions Vehicle_Type  
0  Fine no high winds          Car

In [28]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                   4520
Urban_or_Rural_Area            0
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [29]:
accident['Road_Type'] = accident['Road_Type'].fillna(accident['Road_Type'].mode()[0])

In [30]:
accident.mode()

Index Accident_Severity Accident Date   Latitude Light_Conditions  \
0  2.01E+12            Slight    30/11/2019  52.458798         Daylight   

  District Area  Longitude  Number_of_Casualties  Number_of_Vehicles  \
0    Birmingham  -0.977611                     1                   2   

  Road_Surface_Conditions           Road_Type Urban_or_Rural_Area  \
0                     Dry  Single carriageway               Urban   

   Weather_Conditions Vehicle_Type  
0  Fine no high winds          Car

In [31]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                      0
Urban_or_Rural_Area            0
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [32]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna(accident['Weather_Conditions'].mode()[0])

In [33]:
accident.mode()

Index Accident_Severity Accident Date   Latitude Light_Conditions  \
0  2.01E+12            Slight    30/11/2019  52.458798         Daylight   

  District Area  Longitude  Number_of_Casualties  Number_of_Vehicles  \
0    Birmingham  -0.977611                     1                   2   

  Road_Surface_Conditions           Road_Type Urban_or_Rural_Area  \
0                     Dry  Single carriageway               Urban   

   Weather_Conditions Vehicle_Type  
0  Fine no high winds          Car

In [34]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1>Converting Object to DataTime Data type </h1>

In [35]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce') 

In [36]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [37]:
accident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                       object
dtype: object

<h1>Categorical Data fields</h1>

In [38]:
accident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                       object
dtype: object

In [39]:
accident['Latitude'] = accident['Latitude'].astype('category')
accident['Longitude'] = accident['Longitude'].astype('category')
accident['District Area'] = accident['District Area'].astype('category')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')


In [40]:
accident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                         category
Light_Conditions                 category
District Area                    category
Longitude                        category
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

<h1>Extracting date Information using Pandas Date time</h1>

In [41]:
accident['year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayOfWeek'] = accident['Accident Date'].dt.dayofweek

In [42]:
accident['year']

0         2019
1         2019
2         2019
3         2019
4         2019
          ... 
660674    2022
660675    2022
660676    2022
660677    2022
660678    2022
Name: year, Length: 660679, dtype: int32

<h1>Accident in Light Conditions</h1>

In [43]:
yr19 = accident[accident['year'] == 2019]
yr20 = accident[accident['year'] == 2020]
yr21 = accident[accident['year'] == 2021]
yr22 = accident[accident['year'] == 2022]

In [44]:
yr19['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       133042
Darkness - lights lit           35769
Darkness - no lighting          10842
Darkness - lighting unknown      1768
Darkness - lights unlit           694
Name: count, dtype: int64

In [45]:
yr19['Light_Conditions'].value_counts().sum()

np.int64(182115)

In [46]:
yr20['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       124552
Darkness - lights lit           33473
Darkness - no lighting          10067
Darkness - lighting unknown      1792
Darkness - lights unlit           707
Name: count, dtype: int64

In [47]:
yr20['Light_Conditions'].value_counts().sum()

np.int64(170591)

In [48]:
yr21['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       119518
Darkness - lights lit           32549
Darkness - no lighting           9226
Darkness - lighting unknown      1607
Darkness - lights unlit           654
Name: count, dtype: int64

In [49]:
yr21['Light_Conditions'].value_counts().sum()

np.int64(163554)

In [50]:
yr22['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       107768
Darkness - lights lit           27544
Darkness - no lighting           7302
Darkness - lighting unknown      1317
Darkness - lights unlit           488
Name: count, dtype: int64

In [51]:
yr22['Light_Conditions'].value_counts().sum()


np.int64(144419)

<h1>INSIGHTS:</h1>
<ul>
    <li>Daylight(most number of accident in Light Conditions)2019-2022</li>
    <li>Darkness -light unlit(least number of accident)2019-2022</li>
</ul>

In [52]:
accident['Accident Date'].dt.year.value_counts()

Accident Date
2019    182115
2020    170591
2021    163554
2022    144419
Name: count, dtype: int64

<h1>INSIGHTS 1:</h1>
<ul>
    <li>2019 is the year with the highest number of accident, while 2022, has the lowest number of accident</li>
</ul>

In [53]:
accident['Accident Date'].dt.year.value_counts().sum()

np.int64(660679)

<h1>INSIGHTS2:(naka roundoff na po)</h1>
<ul>
    <li>27.56% of total number of accident in Year 2019</li>
    <li>25.82% of total number of accident in Year 2020</li>
    <li>24.76% of total number of accident in Year 2021</li>
    <li>21.86% of total number of accident in Year 2022</li>
</ul>


<h1>What month has the highest number of accident</h1>

In [54]:
accident['Accident Date'].dt.month.value_counts()

Accident Date
11    60424
10    59580
7     57445
6     56481
9     56455
5     56352
3     54086
8     53913
1     52872
12    51836
4     51744
2     49491
Name: count, dtype: int64

<h1>INSIGHTS3:</h1>
<ul>
    <li>November is the month with the highest number of accident</li>
</ul>

<h1>What type of Vehicle are proned to accident</h1>

In [55]:
accident['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h2>Insights 4: Car is prone to accident</h2>

<h1>Percentage in Accident Severity </h1>

In [56]:
accident['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

In [57]:
accident['Accident_Severity'].value_counts().sum()

np.int64(660679)

<h1>INSIGHTS 5:</h1>
<h2>85.34% of the accident are in Slight Conditions</h2>
<h2>13.35% of the accident are in Serious Conditions</h2>
<h2> 1.31% of the accident are in Serious Conditions</h2>

<h1>INSIGHT 6 </h1>
<h2> Accident severity in Kensington and Chelsea Area </h2>

In [58]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  object        
 1   Accident_Severity        660679 non-null  category      
 2   Accident Date            660679 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  category      
 4   Light_Conditions         660679 non-null  category      
 5   District Area            660679 non-null  category      
 6   Longitude                660679 non-null  category      
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  category      
 10  Road_Type                660679 non-null  category      
 11  Urban_or_Rural_Area      660679 non-null  category      
 12  Weather_Conditio

In [59]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions           District Area Longitude  \
0             Darkness - lights lit  Kensington and Chelsea -0.209082   
1                          Daylight  Kensington and Chelsea -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea -0.210215   
3                          Daylight  Kensington and Chelsea -0.202731   
4             Darkness - lights lit  Kensington and Chelsea -0.192487   
...                             ...                     ...       ...   
660674                     Daylight                Highland -3.467828   
660675       Darkness - no lighting                Highland -3.809281   
660676                     Daylight                Highland -3.862727   
660677       Darkness - no lighting                Highland -3.823997   
660678                     Daylight                Highland -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  year  Month  Day  DayOfWeek  
0                                        Car  2019      6    5          2  
1                                        Car  2019      7    2          1  
2                      Taxi/Private hire car  2019      8   26          0  
3       Bus or coach (17 or more pass seats)  2019      8   16          4  
4                              Other vehicle  2019      9    3          1  
...                                      ...   ...    ...  ...        ...  


In [60]:
accident[(accident['Accident_Severity'] == 'Serious') & (accident['District Area'] == 'Kensington and Chelsea')].count()

Index                      370
Accident_Severity          370
Accident Date              370
Latitude                   370
Light_Conditions           370
District Area              370
Longitude                  370
Number_of_Casualties       370
Number_of_Vehicles         370
Road_Surface_Conditions    370
Road_Type                  370
Urban_or_Rural_Area        370
Weather_Conditions         370
Vehicle_Type               370
year                       370
Month                      370
Day                        370
DayOfWeek                  370
dtype: int64

In [61]:
accident[(accident['Accident_Severity'] == 'Slight') & (accident['District Area'] == 'Kensington and Chelsea')].count()

Index                      2412
Accident_Severity          2412
Accident Date              2412
Latitude                   2412
Light_Conditions           2412
District Area              2412
Longitude                  2412
Number_of_Casualties       2412
Number_of_Vehicles         2412
Road_Surface_Conditions    2412
Road_Type                  2412
Urban_or_Rural_Area        2412
Weather_Conditions         2412
Vehicle_Type               2412
year                       2412
Month                      2412
Day                        2412
DayOfWeek                  2412
dtype: int64

In [62]:
accident[(accident['Accident_Severity'] == 'Fatal') & (accident['District Area'] == 'Kensington and Chelsea')].count()

Index                      17
Accident_Severity          17
Accident Date              17
Latitude                   17
Light_Conditions           17
District Area              17
Longitude                  17
Number_of_Casualties       17
Number_of_Vehicles         17
Road_Surface_Conditions    17
Road_Type                  17
Urban_or_Rural_Area        17
Weather_Conditions         17
Vehicle_Type               17
year                       17
Month                      17
Day                        17
DayOfWeek                  17
dtype: int64

<h1>INSIGHT 7</h1>
<h2> Accident Severity in HighLand Area</h2>

In [63]:
accident[(accident['Accident_Severity'] == 'Slight') & (accident['District Area'] == 'Highland')].count()

Index                      1589
Accident_Severity          1589
Accident Date              1589
Latitude                   1589
Light_Conditions           1589
District Area              1589
Longitude                  1589
Number_of_Casualties       1589
Number_of_Vehicles         1589
Road_Surface_Conditions    1589
Road_Type                  1589
Urban_or_Rural_Area        1589
Weather_Conditions         1589
Vehicle_Type               1589
year                       1589
Month                      1589
Day                        1589
DayOfWeek                  1589
dtype: int64

In [64]:
accident[(accident['Accident_Severity'] == 'Fatal') & (accident['District Area'] == 'Highland')].count()

Index                      88
Accident_Severity          88
Accident Date              88
Latitude                   88
Light_Conditions           88
District Area              88
Longitude                  88
Number_of_Casualties       88
Number_of_Vehicles         88
Road_Surface_Conditions    88
Road_Type                  88
Urban_or_Rural_Area        88
Weather_Conditions         88
Vehicle_Type               88
year                       88
Month                      88
Day                        88
DayOfWeek                  88
dtype: int64

In [65]:
accident[(accident['Accident_Severity'] == 'Serious') & (accident['District Area'] == 'Highland')].count()

Index                      344
Accident_Severity          344
Accident Date              344
Latitude                   344
Light_Conditions           344
District Area              344
Longitude                  344
Number_of_Casualties       344
Number_of_Vehicles         344
Road_Surface_Conditions    344
Road_Type                  344
Urban_or_Rural_Area        344
Weather_Conditions         344
Vehicle_Type               344
year                       344
Month                      344
Day                        344
DayOfWeek                  344
dtype: int64

In [66]:
accident['District Area'].value_counts()

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

In [67]:
accident[(accident['Accident_Severity'] == 'Serious') & (accident['District Area'] == 'Birmingham')].count()

Index                      1474
Accident_Severity          1474
Accident Date              1474
Latitude                   1474
Light_Conditions           1474
District Area              1474
Longitude                  1474
Number_of_Casualties       1474
Number_of_Vehicles         1474
Road_Surface_Conditions    1474
Road_Type                  1474
Urban_or_Rural_Area        1474
Weather_Conditions         1474
Vehicle_Type               1474
year                       1474
Month                      1474
Day                        1474
DayOfWeek                  1474
dtype: int64

In [68]:
accident[(accident['Accident_Severity'] == 'Fatal') & (accident['District Area'] == 'Birmingham')].count()

Index                      105
Accident_Severity          105
Accident Date              105
Latitude                   105
Light_Conditions           105
District Area              105
Longitude                  105
Number_of_Casualties       105
Number_of_Vehicles         105
Road_Surface_Conditions    105
Road_Type                  105
Urban_or_Rural_Area        105
Weather_Conditions         105
Vehicle_Type               105
year                       105
Month                      105
Day                        105
DayOfWeek                  105
dtype: int64

In [69]:
accident[(accident['Accident_Severity'] == 'Slight') & (accident['District Area'] == 'Birmingham')].count()

Index                      11912
Accident_Severity          11912
Accident Date              11912
Latitude                   11912
Light_Conditions           11912
District Area              11912
Longitude                  11912
Number_of_Casualties       11912
Number_of_Vehicles         11912
Road_Surface_Conditions    11912
Road_Type                  11912
Urban_or_Rural_Area        11912
Weather_Conditions         11912
Vehicle_Type               11912
year                       11912
Month                      11912
Day                        11912
DayOfWeek                  11912
dtype: int64

<h1>What is the most common type in Road Surface Conditions </h1>

In [70]:
road_sur_con = accident['Road_Surface_Conditions'].mode()[0]

In [71]:
road_sur_con

'Dry'

<h2>Insight8: The most common Road Surface Conditions is the Dry conditions</h2>

<h1>Number of Casualties in year 2019</h1>

In [72]:
yr19 = accident[accident['year'] == 2019]

In [73]:
yr19

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
182110       2.01E+12            Slight    2019-12-20  54.985289   
182111       2.01E+12           Serious    2019-12-21  54.984105   
182112       2.01E+12            Slight    2019-12-23  55.166369   
182113       2.01E+12            Slight    2019-12-23  54.995154   
182114       2.01E+12           Serious    2019-12-22  55.104618   

                   Light_Conditions           District Area Longitude  \
0             Darkness - lights lit  Kensington and Chelsea -0.209082   
1                          Daylight  Kensington and Chelsea -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea -0.210215   
3                          Daylight  Kensington and Chelsea -0.202731   
4             Darkness - lights lit  Kensington and Chelsea -0.192487   
...                             ...                     ...       ...   
182110       Darkness - no lighting   Dumfries and Galloway -3.210294   
182111                     Daylight   Dumfries and Galloway -3.193693   
182112       Darkness - no lighting   Dumfries and Galloway -2.992068   
182113        Darkness - lights lit   Dumfries and Galloway -3.058338   
182114                     Daylight   Dumfries and Galloway -3.348857   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
182110                     1                   1            Frost or ice   
182111                     2                   1            Frost or ice   
182112                     1                   1            Frost or ice   
182113                     1                   1             Wet or damp   
182114                     2                   2            Frost or ice   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
182110  Single carriageway               Rural                  Other   
182111  Single carriageway               Rural                  Other   
182112  Single carriageway               Rural            Fog or mist   
182113  Single carriageway               Rural     Fine no high winds   
182114  Single carriageway               Rural  Raining no high winds   

                                Vehicle_Type  year  Month  Day  DayOfWeek  
0                                        Car  2019      6    5          2  
1                                        Car  2019      7    2          1  
2                      Taxi/Private hire car  2019      8   26          0  
3       Bus or coach (17 or more pass seats)  2019      8   16          4  
4                              Other vehicle  2019      9    3          1  
...                                      ...   ...    ...  ...        ...  


In [74]:
yr19['Number_of_Casualties'].value_counts().sum()

np.int64(182115)

<h2>Insight 9: There are 182,155 casualties in total, year 2019</h2>

<h1>Number of Casualties in year 2020</h1>

In [75]:
yr20 = accident[accident['year'] == 2020]

In [76]:
yr20['Number_of_Casualties'].value_counts().sum()

np.int64(170591)

<h2>Insight10: There are 170,591 casualties in year 2020</h2>

<h1>Number of Casualties in year 2021</h1>

In [77]:
yr21 = accident[accident['year'] == 2021]

In [78]:
yr21['Number_of_Casualties'].value_counts().sum()

np.int64(163554)

<h2>Insight11: There are 163,554 casualties in year 2021</h2>

<h1>Number of Casualties in year 2022</h1>

In [79]:
yr22 = accident[accident['year'] == 2022] 

In [80]:
yr22['Number_of_Casualties'].value_counts().sum()

np.int64(144419)

<h2>Insight12: There are 144,419 casualties in year 2022</h2>

<h1>How many accident happened in Rural_Urban area (2019)</h1>

In [81]:
yr19 = accident[accident['year'] == 2019]
yr19['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          114552
Rural           67558
Unallocated         5
Name: count, dtype: int64

<h2>Insight 13: In 2019, Most accident happened in Urban Area  </h2>

<h1>How many accident happened in Rural_Urban area (2020)</h1>

In [82]:
yr20 = accident[accident['year'] == 2020]
yr20['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          108594
Rural           61991
Unallocated         6
Name: count, dtype: int64

<h2>Insight 14: Just like the insight 8, but in year 2020. Most accident happened in Urban Area </h2>

<h1>How many accident happened in Rural_Urban area (2021)</h1>

In [83]:
yr21 = accident[accident['year'] == 2021]
yr21['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          104538
Rural           59016
Unallocated         0
Name: count, dtype: int64

<h2>Insight 15: There are 104538 accident happened in Urban area while in Rural are 59016</h2>

<h1>How many accident happened in Rural_Urban area (2022)</h1>

In [84]:
yr22 = accident[accident['year'] == 2022]
yr22['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          93994
Rural          50425
Unallocated        0
Name: count, dtype: int64

<h2>Insight 16: There are 93,994 accident happened in Urban area while in Rural are 50425</h2>

<h1> INSIGHT 12</h1>

In [85]:
accident_insight = accident.groupby(['Number_of_Casualties', 'Accident Date']).size()

In [86]:
accident_insight

Number_of_Casualties  Accident Date
1                     2019-01-01       240
                      2019-01-02       284
                      2019-01-03       294
                      2019-01-04       306
                      2019-01-05       342
                                      ... 
45                    2020-09-26         1
47                    2020-07-17         1
48                    2021-12-22         1
62                    2020-08-18         1
68                    2019-01-03         1
Length: 9056, dtype: int64

<h1> Insight 17: Most common type of vehicle prone to accident</h1>

In [87]:
accident['Vehicle_Type'].mode()[0]

'Car'

<h1>What is the common Weather Conditions</h1>

In [88]:
weather_con = accident['Weather_Conditions'].mode()[0]

In [89]:
weather_con

'Fine no high winds'

<h2>Insight 18: The most common weather conditions is fine no high winds </h2>

<h1>Accident by District Area</h1>

In [90]:
accidentbyDistrict = accident['District Area'].value_counts()

In [91]:
accidentbyDistrict

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h2>Insight 19: BIRMINGHAM the most number of accident </h2>

<h1> what is the total number of Casualties in All accident?</h1>

In [92]:
sum_casualties = accident['Number_of_Casualties'].sum()

In [93]:
sum_casualties

np.int64(896568)

<h2>Insight 20: The total number of Casualties happened in all accident is around 896,568 </h2>

<h1> What is the total Number_of_Vehicles in All accident?</h1>

In [94]:
sum_vehicles = accident['Number_of_Vehicles'].sum()

In [95]:
sum_vehicles

np.int64(1209872)

<h2></h2>

<h1> Which District Area have the most number of Casualties</h1>

In [96]:
casualties_district = accident.groupby('District Area')['Number_of_Casualties'].sum()

C:\Users\Light Micro Finance\AppData\Local\Temp\ipykernel_10872\1293022102.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  casualties_district = accident.groupby('District Area')['Number_of_Casualties'].sum()


In [97]:
casualties_district

District Area
Aberdeen City    1508
Aberdeenshire    2529
Adur              817
Allerdale        1663
Alnwick           351
                 ... 
Wychavon         1943
Wycombe          2442
Wyre             1706
Wyre Forest      1340
York             2453
Name: Number_of_Casualties, Length: 422, dtype: int64

<h2>Insight 21: So the Aberdeen City has the most number of Casualties</h2>

<h1> what vehicle type are proned to accident</h1>

In [98]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  object        
 1   Accident_Severity        660679 non-null  category      
 2   Accident Date            660679 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  category      
 4   Light_Conditions         660679 non-null  category      
 5   District Area            660679 non-null  category      
 6   Longitude                660679 non-null  category      
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  category      
 10  Road_Type                660679 non-null  category      
 11  Urban_or_Rural_Area      660679 non-null  category      
 12  Weather_Conditio

In [104]:
accident.groupby('Vehicle_Type')['Accident_Severity'].value_counts().sum()

C:\Users\Light Micro Finance\AppData\Local\Temp\ipykernel_10872\8843315.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  accident.groupby('Vehicle_Type')['Accident_Severity'].value_counts().sum()


np.int64(660679)

In [105]:
accident['Vehicle_Type']

0                                          Car
1                                          Car
2                        Taxi/Private hire car
3         Bus or coach (17 or more pass seats)
4                                Other vehicle
                          ...                 
660674                                     Car
660675                                     Car
660676                                     Car
660677                   Motorcycle over 500cc
660678                                     Car
Name: Vehicle_Type, Length: 660679, dtype: category
Categories (16, object): ['Agricultural vehicle', 'Bus or coach (17 or more pass seats)', 'Car', 'Data missing or out of range', ..., 'Pedal cycle', 'Ridden horse', 'Taxi/Private hire car', 'Van / Goods 3.5 tonnes mgw or under']

<h1>Insight 22: The vehicle type that is mostly prone to accidents is the Car</h1>

In [100]:
casualties_roadWeather = accident.groupby(['Road_Type', 'Weather_Conditions'])['Number_of_Casualties'].sum().unstack()

C:\Users\Light Micro Finance\AppData\Local\Temp\ipykernel_10872\807559471.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  casualties_roadWeather = accident.groupby(['Road_Type', 'Weather_Conditions'])['Number_of_Casualties'].sum().unstack()


In [101]:
casualties_roadWeather

Weather_Conditions  Fine + high winds  Fine no high winds  Fog or mist  Other  \
Road_Type                                                                       
Dual carriageway                 2191              116022         1136   3246   
One way street                    187               13482           36    381   
Roundabout                        720               45619          293   1408   
Single carriageway               8610              537717         3615  17939   
Slip road                         148                8035           46    262   

Weather_Conditions  Raining + high winds  Raining no high winds  \
Road_Type                                                         
Dual carriageway                    3000                  19598   
One way street                       200                   1775   
Roundabout                           706                   6911   
Single carriageway                  9562                  82663   
Slip road                            153                   1282   

Weather_Conditions  Snowing + high winds  Snowing no high winds  
Road_Type                                                        
Dual carriageway                     257                   1427  
One way street                        15                     96  
Roundabout                            46                    382  
Single carriageway                   924                   6380  
Slip road                             13                     85

In [102]:
percentage_weather = casualties_roadWeather.div(casualties_roadWeather.sum().sum())*100

In [103]:
percentage_weather

Weather_Conditions  Fine + high winds  Fine no high winds  Fog or mist  \
Road_Type                                                                
Dual carriageway             0.244376           12.940680     0.126705   
One way street               0.020857            1.503734     0.004015   
Roundabout                   0.080306            5.088181     0.032680   
Single carriageway           0.960329           59.975038     0.403204   
Slip road                    0.016507            0.896195     0.005131   

Weather_Conditions     Other  Raining + high winds  Raining no high winds  \
Road_Type                                                                   
Dual carriageway    0.362047              0.334609               2.185891   
One way street      0.042495              0.022307               0.197977   
Roundabout          0.157043              0.078745               0.770828   
Single carriageway  2.000852              1.066511               9.219936   
Slip road           0.029223              0.017065               0.142990   

Weather_Conditions  Snowing + high winds  Snowing no high winds  
Road_Type                                                        
Dual carriageway                0.028665               0.159162  
One way street                  0.001673               0.010707  
Roundabout                      0.005131               0.042607  
Single carriageway              0.103060               0.711602  
Slip road                       0.001450               0.009481